### ISSUE TITLE
OOM killed after processing phase

### ISSUE THREAD

**Issue by pnorman** [openstreetmap/osm2pgsql#351]
On a machine with 32GB RAM, osm2pgsql 0.87.3, command line `~/osm2pgsql/osm2pgsql -d gis --cache 26000 --flat-nodes nodes.bin --slim --drop --
number-processes 8 -S ~/openstreetmap-carto/openstreetmap-carto.style planet-150202.osm.pbf`

```
osm2pgsql SVN version 0.87.3 (64bit id space)

Using built-in tag processing pipeline
Using projection SRS 900913 (Spherical Mercator)
Setting up table: planet_osm_point
Setting up table: planet_osm_line
Setting up table: planet_osm_polygon
Setting up table: planet_osm_roads
Allocating memory for dense node cache
Allocating dense node cache in one big chunk
Allocating memory for sparse node cache
Sharing dense sparse
Node-cache: cache=26000MB, maxblocks=3328000*8192, allocation method=11
Mid: loading persistent node cache from nodes.bin
Allocated space for persistent node cache file
Maximum node in persistent node cache: 0
Mid: pgsql, scale=100 cache=26000
Setting up table: planet_osm_nodes
Setting up table: planet_osm_ways
Setting up table: planet_osm_rels

Reading in file: planet-150202.osm.pbf
Processing: Node(2715748k 2596.3k/s) Way(272021k 51.88k/s) Relation(2630170 269.21/s)
Standard exception processing relation id=3890452: TopologyException: side location conflict at 1038877.48 5779728.6200000001
Processing: Node(2715748k 2596.3k/s) Way(272021k 51.88k/s) Relation(2841950 273.71/s)
Standard exception processing relation id=4136486: TopologyException: side location conflict at 2924456.4100000001 9012036.7799999993
Processing: Node(2715748k 2596.3k/s) Way(272021k 51.88k/s) Relation(2881850 275.04/s)
Standard exception processing relation id=4180656: TopologyException: side location conflict at 272703.48999999999 5093940.6600000001
Processing: Node(2715748k 2596.3k/s) Way(272021k 51.88k/s) Relation(3100780 281.58/s)
Standard exception processing relation id=4542207: TopologyException: side location conflict at 15042893.51 4046892.75
Processing: Node(2715748k 2596.3k/s) Way(272021k 51.88k/s) Relation(3109640 282.05/s)  parse time: 17314s
Node stats: total(2715748725), max(3326469484) in 1046s
Way stats: total(272021914), max(326028236) in 5243s
Relation stats: total(3109645), max(4551629) in 11025s
Maximum node in persistent node cache: 3327131647
Committing transaction for planet_osm_point
Committing transaction for planet_osm_line
Committing transaction for planet_osm_polygon
Committing transaction for planet_osm_roads
Command terminated by signal 9
```

Watching `status` in /proc/PID tells me that about 24500 cache would have been sufficient.

There are about 8GB free as it finishes processing

![image](https://cloud.githubusercontent.com/assets/1190866/7776708/de58f4b2-0071-11e5-8566-602376c3f0f1.png)

I haven't yet looked through the code to see what could be using large amounts of memory. `vm.overcommit_memory` is set to 1, so an allocation alone won't cause a problem, only use. `vm.overcommit_ratio` is 50, but I don't believe that matters with overcommit_memory set to 1.

**Comment by pnorman** [comment]
I ran with `--cache 12000` to see what happens, and something is taking up gobs of memory in pending

![image](https://cloud.githubusercontent.com/assets/1190866/7781861/b4da7bbe-00ae-11e5-80e5-374d03edad24.png)

**Comment by pnorman** [comment]
after #353, or more precisely something in [0.87.3...master](https://github.com/openstreetmap/osm2pgsql/compare/0.87.3...master), it is worse, running out of memory in nodes processing.

![image](https://cloud.githubusercontent.com/assets/1190866/7804108/e95e163c-0312-11e5-8bb2-9f13d955bf9d.png)


This is with `--cache 12000``, so memory usage shouldn't be much above the initial spike of committed to about 13G

**Comment by lonvia** [comment]
I've reverted the part that is causing the second leak. No lead as to what is actually using the memory yet.

**Comment by lonvia** [comment]
There is quite a bit of memory allocation going on before processing of pending. The memory cache for the flatnode store alone allocates about 650MB (80MB per thread).

The most likely candidate for the jump in allocations above however is the pending queue. All pending ways are initially put in the queue with each entry taking up 16 byte (an osmid and a pointer). I haven't tested with a full planet but for extracts about 60% of the ways need to be reprocessed, so for the planet about 160M ways. That leaves the queue with a total of 2.5G of memory. Assuming that queue behaves like a vector which doubles it's size during each reallocation, the queue might easily take up 5G of memory (make that 7.5GB shortly during reallocation when the vector is expanded).

It should be possible to avoid the allocation of cache memory for the flatnode store when there is enough ram cache. The pending queue issue is more difficult to fix. A size-bound queue where middle acts as a producer is probably the most reasonable solution but not exactly easy to implement. (And just to mention it, there is of course also the solution to drop support for old-style multipolygons. Then we could just get rid of all this processing of pending ways and get a huge performance improvement on the side.)

@pnorman Any idea how to confirm the theory with the pending queue on a planet import?

**Comment by pnorman** [comment]
> The most likely candidate for the jump in allocations above however is the pending queue. All pending ways are initially put in the queue with each entry taking up 16 byte (an osmid and a pointer). I haven't tested with a full planet but for extracts about 60% of the ways need to be reprocessed, so for the planet about 160M ways. That leaves the queue with a total of 2.5G of memory. Assuming that queue behaves like a vector which doubles it's size during each reallocation, the queue might easily take up 5G of memory (make that 7.5GB shortly during reallocation when the vector is expanded).

So in processing we're building them up into a different data structure, then in pending, copying from that to a less space-efficient queue?

**Comment by lonvia** [comment]
> So in processing we're building them up into a different data structure, then in pending, copying from that to a less space-efficient queue?

Jup, that's what I read from in the code.

**Comment by pnorman** [comment]
> > So in processing we're building them up into a different data structure, then in pending, copying from that to a less space-efficient queue?
> Jup, that's what I read from in the code.

Two thoughts then

1. Read directly from the initial data structure instead of the queue. That would mean making the initial data structure thread-safe and efficient at reading from it in parallel. Or having each thread only read ways that have way_id mod num-threads = thread number (or similar logic working on groups of 1k ways)

2. Copying into the queue in chunks, copying in a new chunk whenever it gets empty.

**Comment by pnorman** [comment]
I tried --without-lockfree, and the RAM usage is substantially better

![image](https://cloud.githubusercontent.com/assets/1190866/8447581/d7f7cb8a-1f6f-11e5-9669-3694f6876498.png)

So, I guess the options are

1. Fix the boost::lockfree::queue implementation to use less memory
2. Switch the configure logic to default to --without-lockfree
3. Drop the boost::lockfree::queue implementation and simplify ./configure

**Comment by pnorman** [comment]
cc @kevinkreiser

When porting to C++ I found the lockfree queue did help, but I don't think this was on a full planet, so memory usage was not an issue.

On a typical machine, the difference between all nodes in memory cache and having to hit disk is the biggest cause of performance differences outside of end of import indexing/clustering. On one with mechanical hard drives, this is even more so. So, besides being confusing and annoying, this bug is the biggest performance issue currently present.

On the machine I use for benchmarking, this is probably less so, as it has two SSDs in RAID0.

I'm running a benchmark to verify this.

**Comment by lonvia** [comment]
Looks like only the 2.5G from my back-of-the-envelope calculation are now left. I'd say that is good enough and lean towards option 3 as it simplifies the code as well.

**Comment by zbycz** [comment]
Just for other users reference, I have the same issue with geofabrics europe.pbf (16G, 4cpu, SSD:120G).

```
$ osm2pgsql -C 8000 --slim --flat-nodes /second/nodes.flat --number-processes 4 /second/europe-latest.osm.pbf

Processing: Node(1526618k 684.0k/s) Way(186582k 7.77k/s) Relation(2567130 57.96/s)  parse time: 70520s
Node stats: total(1526618933), max(3662282141) in 2232s
Way stats: total(186582734), max(361799115) in 23999s
Relation stats: total(2567138), max(5385560) in 44288s
Committing transaction for planet_osm_point
Committing transaction for planet_osm_line
Committing transaction for planet_osm_polygon
Committing transaction for planet_osm_roads

Going over pending ways...
Maximum node in persistent node cache: 3662675967
Killed

```

**Comment by andrew-aladev** [comment]
@pnorman This is still an issue. You've created a manual http://wiki.openstreetmap.org/wiki/Osm2pgsql/benchmarks. This manual is invalid. I've used osm2pgsql with `-C 16000` 30 GB ram, 30 GB swap and 800 GB SSD. It has been killed with OOM.

**Comment by lonvia** [comment]
@andrew-aladev This issue refers to a two year old version of osm2pgsql and the OOM happens at a completely different point during the import, so it is save to say that your problem is completely unrelated. 

Please open a new issue, state which version and OS you were using, what command line and which data you were importing.

**Comment by andrew-aladev** [comment]
ok
